File: modules/angular2/src/web-workers/ui/application.dart
library angular2.src.web_workers.ui;

import 'dart:isolate';
import 'dart:async';
import 'dart:core';
import 'package:angular2/src/web-workers/shared/message_bus.dart'
    show MessageBus, MessageBusSink, MessageBusSource;
import 'package:angular2/src/web-workers/ui/impl.dart' show bootstrapUICommon;

/**
 * Bootstrapping a WebWorker
 *
 * You instantiate a WebWorker application by calling bootstrap with the URI of your worker's index script
 * Note: The WebWorker script must call bootstrapWebworker once it is set up to complete the bootstrapping process
 */
Future<MessageBus> bootstrap(String uri) {
  return spawnWebWorker(Uri.parse(uri)).then((bus) {
    bootstrapUICommon(bus);
    return bus;
  });
}

/**
 * To be called from the main thread to spawn and communicate with the worker thread
 */
Future<UIMessageBus> spawnWebWorker(Uri uri) {
  var receivePort = new ReceivePort();
  var isolateEndSendPort = receivePort.sendPort;
  return Isolate.spawnUri(uri, const [], isolateEndSendPort).then((_) {
    var source = new UIMessageBusSource(receivePort);
    return source.sink.then((sendPort) {
      var sink = new UIMessageBusSink(sendPort);
      return new UIMessageBus(sink, source);
    });
  });
}

class UIMessageBus extends MessageBus {
  final UIMessageBusSink sink;
  final UIMessageBusSource source;

  UIMessageBus(UIMessageBusSink sink, UIMessageBusSource source)
      : sink = sink,
        source = source;
}

class UIMessageBusSink extends MessageBusSink {
  final SendPort _port;

  UIMessageBusSink(SendPort port) : _port = port;

  void send(message) {
    _port.send(message);
  }
}

class UIMessageBusSource extends MessageBusSource {
  final ReceivePort _port;
  final Stream rawDataStream;
  Map<int, StreamSubscription> _listenerStore =
      new Map<int, StreamSubscription>();
  int _numListeners = 0;

  UIMessageBusSource(ReceivePort port)
      : _port = port,
        rawDataStream = port.asBroadcastStream();

  Future<SendPort> get sink => rawDataStream.firstWhere((message) {
        return message is SendPort;
      });

  int addListener(Function fn) {
    var subscription = rawDataStream.listen((message) {
      fn({"data": message});
    });

    _listenerStore[++_numListeners] = subscription;
    return _numListeners;
  }

  void removeListener(int index) {
    _listenerStore[index].cancel();
    _listenerStore.remove(index);
  }
}

File: modules/angular2/src/web-workers/ui/application.ts
import {
  MessageBus,
  MessageBusSource,
  MessageBusSink,
  SourceListener
} from "angular2/src/web-workers/shared/message_bus";
import {BaseException} from "angular2/src/facade/lang";
import {bootstrapUICommon} from "angular2/src/web-workers/ui/impl";

/**
 * Bootstrapping a WebWorker
 *
 * You instantiate a WebWorker application by calling bootstrap with the URI of your worker's index
 * script
 * Note: The WebWorker script must call bootstrapWebworker once it is set up to complete the
 * bootstrapping process
 */
export function bootstrap(uri: string): MessageBus {
  var messageBus = spawnWorker(uri);
  bootstrapUICommon(messageBus);
  return messageBus;
}

export function spawnWorker(uri: string): MessageBus {
  var worker: Worker = new Worker(uri);
  return new UIMessageBus(new UIMessageBusSink(worker), new UIMessageBusSource(worker));
}

export class UIMessageBus implements MessageBus {
  constructor(public sink: UIMessageBusSink, public source: UIMessageBusSource) {}
}

export class UIMessageBusSink implements MessageBusSink {
  constructor(private _worker: Worker) {}

  send(message: Object): void { this._worker.postMessage(message); }
}

export class UIMessageBusSource implements MessageBusSource {
  private _listenerStore: Map<int, SourceListener> = new Map<int, SourceListener>();
  private _numListeners: int = 0;

  constructor(private _worker: Worker) {}

  public addListener(fn: SourceListener): int {
    this._worker.addEventListener("message", fn);
    this._listenerStore[++this._numListeners] = fn;
    return this._numListeners;
  }

  public removeListener(index: int): void {
    removeEventListener("message", this._listenerStore[index]);
    this._listenerStore.delete(index);
  }
}
